Added new @slack/socket-mode package (#1096)
slackapi/node-slack-sdk: docs/_packages/socket_mode.md

---
title: Socket Mode
permalink: /socket-mode
order: 7
anchor_links_header: Usage
---

# Slack Socket Mode

## Installation

```shell
$ npm install @slack/socket-mode
```

---

### Initialize the client

> The package is designed for the newly released [**Socket Mode**](https://api.slack.com/socket-mode) feature. It exports a `SocketModeClient` class. Your app will create an instance of the class for each workspace it

Reviewer: ```suggestion
This package is designed to support [**Socket Mode**](https://api.slack.com/socket-mode), which allows your app to receive events from Slack over a WebSocket connection.

The package exports a `SocketModeClient` class. Your app will create an instance of the class for each workspace it
```

> communicates with. Creating an instance requires an **app level token** from Slack. Apps connect to the **Socket Mode** API using the newly introduced **app level token**, which starts with `xapp`.

Reviewer: Is there a line break between this and the last??? If so, could it be removed?

Reviewer: In the other guide, you use "App Token". Regardless of what is used, could we make it consistent across the guides?

The API site uses "app-level token"

> Note: **Socket Mode** requires the scope `connections:write`. In your browser, navigate to your [app's app config](https://api.slack.com/apps) and go to the **OAuth and Permissions** section to add the scope.

Reviewer: ```suggestion
Note: **Socket Mode** requires the `connections:write` scope. Navigate to your [app configuration](https://api.slack.com/apps) and go to the **OAuth and Permissions** section to add the scope.
```

```javascript
const { SocketModeClient } = require('@slack/socket-mode');

// Read a token from the environment variables
const appToken = process.env.SLACK_APP_TOKEN;

// Initialize
const client = new SocketModeClient({appToken});
```

### Connect to Slack

> Data from Slack will begin to flow to your app once the client is connected. You'll also be able to send data to

Reviewer: ```suggestion
After your client establishes a connection, your app can send data to and receive data from Slack.
```

Reviewer: Weird thing is happening with suggestions, but if you take suggestion above, remove the fragment below (but keep the "Connecting..." sentence)

Slack after the connection is established. Connecting is as easy as calling the `.start()` method.

```javascript
const { SocketModeClient } = require('@slack/socket-mode');
const appToken = process.env.SLACK_APP_TOKEN;

const socketModeClient = new SocketModeClient(appToken);

(async () => {
  // Connect to Slack
  await socketModeClient.start();
})();
```
---

### Listen for an event

> Apps register functions, called listeners, to be triggered when an event of a specific type is received by the client.

Reviewer: ```suggestion
Bolt apps register [listener functions](https://slack.dev/bolt-js/reference#listener-functions), which are triggered when a specific event type is received by the client.
```

If you've used Node's [`EventEmitter`](https://nodejs.org/api/events.html#events_class_eventemitter) pattern
before, then you're already familiar with how this works, since the client is an `EventEmitter`.

The `event` argument passed to the listener is an object. Its content corresponds to the [type of
event](https://api.slack.com/events) it's registered for.

```javascript
const { SocketModeClient } = require('@slack/socket-mode');
const appToken = process.env.SLACK_APP_TOKEN;

const socketModeClient = new SocketModeClient(appToken);

// Attach listeners to events by type. See: https://api.slack.com/events/message
socketModeClient.on('message', (event) => {
  console.log(event);
});

(async () => {
  await socketModeClient.start();
})();
```

---

### Send a message

> To respond to events and send messages back into slack, we recommend using the `@slack/web-api` package with a `bot token`.

Reviewer: ```suggestion
To respond to events and send messages back into Slack, we recommend using the `@slack/web-api` package with a `bot token`.
```

const { WebClient } = require('@slack/web-api');

const socketModeClient = new SocketModeClient(process.env.SLACK_APP_TOKEN);
const webClient = new WebClient(process.env.BOT_TOKEN);

// Attach listeners to events by type. See: https://api.slack.com/events/message
socketModeClient.on('member_joined_channel', async ({event, body, ack}) => {
    try {
      // send acknowledgement back to slack over the socketMode websocket connection
      // this is so slack knows you have received the event and are processing it
      await ack();
      await webClient.chat.postMessage({
          blocks: [
          {
            type: 'section',
            text: {
              type: 'mrkdwn',
              text: `Welcome to the channel, <@${event.user}>. We're here to help. Let us know if you have an issue.`,
            },
            accessory: {
              type: 'button',
              text: {
                type: 'plain_text',
                text: 'Get Help',
              },
              value: 'get_help',
            },
          },
        ],
        channel: event.channel,
      });
    } catch (error) {
      console.log('An error occurred', error);
    }
  });
```
---

> ### Lifecycle events

Reviewer: This section is really solid and well-framed

|-----------------|-----------------|-------------|
| `connecting`    |  | The client is in the process of connecting to the platform. |
| `authenticated` | `(connectData)` - the response from `apps.connections.open` | The client has authenticated with the platform. This is a sub-state of `connecting`. |
| `connected`     |  | The client is connected to the platform and incoming events will start being emitted. |
| `ready`         |  | The client is ready to send outgoing messages. This is a sub-state of `connected` |
| `disconnecting` |  | The client is no longer connected to the platform and cleaning up its resources. It will soon transition to `disconnected`. |
| `reconnecting`  |  | The client is no longer connected to the platform and cleaning up its resources. It will soon transition to `connecting`. |
| `disconnected`  | `(error)` | The client is not connected to the platform. This is a steady state - no attempt to connect is occurring. The `error` argument will be `undefined` when the client initiated the disconnect (normal). |

The client also emits events that are part of its lifecycle, but aren't states. Instead, they represent specific

> moments that might be interesting to your app. The following table lists these events, their description, and includes

Reviewer: ```suggestion
moments that might be helpful to your app. The following table lists these events, their description, and includes
```

the arguments that a listener would receive.

| Event Name      | Arguments | Description |
|-----------------|-----------|-------------|
| `error`         | `(error)` | An error has occurred. See [error handling](#handle-errors) for details. |
| `slack_event`   | `(eventType, event)` | An incoming Slack event has been received. |
| `unable_to_socket_mode_start` | `(error)` | A problem occurred while connecting, a reconnect may or may not occur. |

---

### Logging

The `SocketModeClient` will log interesting information to the console by default. You can use the `logLevel` to decide how

> much information, or how interesting the information needs to be, in order for it to be output. There are a few possible

Reviewer: ```suggestion
much or what kind of information should be output. There are a few possible
```

---
title: Socket Mode
permalink: /socket-mode
order: 7
anchor_links_header: Usage
---

# Slack Socket Mode

## Installation

```shell
$ npm install @slack/socket-mode
```

---

### Initialize the client

The package exports an `SocketModeClient` class. Your app will create an instance of the class for each workspace it
communicates with. Creating an instance requires a `app level token` from Slack. Apps connect to the `Socket Mode` API using the newly introduced `app level token`, which start with `xapp`.

Note: `Socket Mode` requires the scope of `connections:write`. In your browser, navigate to your app's app config and go to the `OAuth and Permissions` section to add the scope.

```javascript
const { SocketModeClient } = require('@slack/socket-mode');

// Read a token from the environment variables
const appToken = process.env.SLACK_APP_TOKEN;

// Initialize
const client = new SocketModeClient({appToken});
```

### Connect to Slack

> Data from Slack will begin to flow to your program once the client is connected. You'll also be able to send data to

Reviewer 2: ```suggestion
Data from Slack will begin to flow to your app once the client is connected. You'll also be able to send data to
```

Slack after the connection is established. Connecting is as easy as calling the `.start()` method.

```javascript
const { SocketModeClient } = require('@slack/socket-mode');
const appToken = process.env.SLACK_APP_TOKEN;

const socketModeClient = new SocketModeClient(appToken);

(async () => {
  // Connect to Slack
  await socketModeClient.start();
})();
```
---

### Listen for an event

Apps register functions, called listeners, to be triggered when an event of a specific type is received by the client.
If you've used Node's [`EventEmitter`](https://nodejs.org/api/events.html#events_class_eventemitter) pattern
before, then you're already familiar with how this works, since the client is an `EventEmitter`.

> The `event` argument passed to the listener is an object. It's contents corresponds to the [type of

Reviewer 2: ```suggestion
The `event` argument passed to the listener is an object. Its content corresponds to the [type of
```

> event](https://api.slack.com/events) its registered for.

Reviewer 2: ```suggestion
event](https://api.slack.com/events) it's registered for.
```

```javascript
const { SocketModeClient } = require('@slack/socket-mode');
const appToken = process.env.SLACK_APP_TOKEN;

const socketModeClient = new SocketModeClient(appToken);

// Attach listeners to events by type. See: https://api.slack.com/events/message
socketModeClient.on('message', (event) => {
  console.log(event);
});

(async () => {
  await socketModeClient.start();
})();
```

---

### Send a message

> To respond to events and send messages back into slack, it is recommend to use the `@slack/web-api` package with a `bot token`.

Reviewer 2: ```suggestion
To respond to events and send messages back into slack, we recommend using the `@slack/web-api` package with a `bot token`.
```

          blocks: [
          {
            type: 'section',
            text: {
              type: 'mrkdwn',
              text: `Welcome to the channel, <@${event.user}>. We're here to help. Let us know if you have an issue.`,
            },
            accessory: {
              type: 'button',
              text: {
                type: 'plain_text',
                text: 'Get Help',
              },
              value: 'get_help',
            },
          },
        ],
        channel: event.channel,
      });
    } catch (error) {
      console.log('An error occurred', error);
    }
  });
```
---

### Lifecycle events

The client's connection to Slack has a lifecycle. This means the client can be seen as a state machine which transitions
through a few states as it connects, disconnects, reconnects, and synchronizes with Slack. The client emits an event
for each state it transitions to throughout its lifecycle. If your app simply needs to know whether the client is
connected or not, the `.connected` boolean property can be checked.

In the table below, the client's states are listed, which are also the names of the events you can use to observe

> the transition to that state. The table also includes description for the state, and arguments that a listener would

Reviewer 2: ```suggestion
the transition to that state. The table also includes descriptions for the states and arguments that a listener would
```

The client also emits events that are part of its lifecycle, but aren't states. Instead, they represent specific
moments that might be interesting to your app. The following table lists these events, their description, and includes
the arguments that a listener would receive.

| Event Name      | Arguments | Description |
|-----------------|-----------|-------------|
| `error`         | `(error)` | An error has occurred. See [error handling](#handle-errors) for details. |
| `slack_event`   | `(eventType, event)` | An incoming Slack event has been received. |
| `unable_to_socket_mode_start` | `(error)` | A problem occurred while connecting, a reconnect may or may not occur. |

---

### Logging

The `SocketModeClient` will log interesting information to the console by default. You can use the `logLevel` to decide how
much information, or how interesting the information needs to be, in order for it to be output. There are a few possible
log levels, which you can find in the `LogLevel` export. By default, the value is set to `LogLevel.INFO`. While you're

> in development, its sometimes helpful to set this to the most verbose: `LogLevel.DEBUG`.

Reviewer 2: ```suggestion
in development, it's sometimes helpful to set this to the most verbose: `LogLevel.DEBUG`.
```

Reviewer 2: Have we thought about turning on `LogLevel.DEBUG` when in development? For example, when there is no set value for `process.env.NODE_ENV`?

I think this would be helpful to developers and also issues submitted to the Issue tracker.
